python/lsst/ctrl/mpexec/cmdLineFwk.py

```diff
@@ -66,6 +76,148 @@
+            self.exists = True
+        except MissingCollectionError:
+            self.exists = False
+
+    name: str
+    exists: bool
+
+
+@dataclass
+class _ButlerFactory:
+
+    def __init__(self, registry: Registry, args: argparse.Namespace, writeable: bool):
+        if args.output is not None:
+            self.output = _OutputChainedCollectionInfo(registry, args.output)
+        else:
+            self.output = None
+        if args.output_run is not None:
+            self.outputRun = _OutputRunCollectionInfo(registry, args.output_run)
+        elif self.output is not None:
+            if args.extend_run:
+                runName, _ = self.output.chain[0]
+            else:
+                runName = "{}/{:%Y%m%dT%Hh%Mm%Ss}".format(self.output, datetime.datetime.now())
+            self.outputRun = _OutputRunCollectionInfo(registry, runName)
+        elif not writeable:
+            # If we're not writing yet, ok to have no output run.
+            self.outputRun = None
+        else:
+            raise ValueError("Cannot write without at least one of (--output, --output-run).")
+        self.inputs = list(CollectionSearch.fromExpression(args.input))
+
+    def check(self, args: argparse.Namespace):
+        # Check options for consistency with each other and repo.
+        assert not (args.extend_run and args.replace_run), "In mutually-exclusive group in ArgumentParser."
+        if self.inputs and self.output is not None and self.output.exists:
+            raise ValueError("Cannot use --output with existing collection with --inputs.")
+        if args.extend_run and self.outputRun is None:
+            raise ValueError("Cannot --extend-run when no output collection is given.")
+        if args.extend_run and not self.outputRun.exists:
+            raise ValueError(f"Cannot --extend-run; output collection "
+                             f"'{self.outputRun.name}' does not exist.")
+        if not args.extend_run and self.outputRun.exists:
+            raise ValueError(f"Output run '{self.outputRun.name}' already exists, but "
+                             f"--extend-run was not given.")
+        if args.prune_replaced and not args.replace_run:
+            raise ValueError(f"--prune-replaced requires --replace-run.")
+        if args.replace_run and (self.output is None or not self.output.exists):
+            raise ValueError(f"--output must point to an existing CHAINED collection for --replace-run.")
+
+    @classmethod
+    def _makeReadParts(cls, args: argparse.Namespace):
+        butler = Butler(args.butler_config, writeable=False)
+        self = cls(butler.registry, args, writeable=False)
+        self.check(args)
+        if self.output and self.output.exists:
+            if args.replace_run:
+                replaced, _ = self.output.chain[0]
+                inputs = self.output.chain[1:]
+                _LOG.debug(f"Simulating collection search in '{self.output.name}' "
```

Comment:
I think we prefer to defer the string interpolation in log messages (especially for debug) so should not use f-strings.

Reply:
It's hard for me to believe that could possibly ever matter in Python (I think it's a dubious habit we've gotten into because it _does_ make sense in C++), but it is at least precedent.  Will change.